DEVOP-1668 Improve simplified script

otoumas opened the pull request:
In this pull request, the compose project has been fully refactored.

Main changes:

- Refactor of the project structure: new .env template files, new compose files
- Refactor of init script: Now the script is used to setup the project into the deployments directory for each type of node (rpc or forger) and network (eon or gobi). It also provides instructions on how to run the newly created stack.
- seed.sh script improvements: Now the seed script will first check if there is a need to reseed, first looking if the .seed.complete fie exists, then if blocks or chainstate directories exist. The new .env templates provide new env vars to control this script.
- Updated docs

Commit:
Refactor project structure and init scripts.
Add more functionality to seed.sh

lander86 commented on docs/FORGER.md:
```diff
@@ -0,0 +1,115 @@
+--- 
+
+1. Prerequisites
+    - Storage: A minimum of 250GB of free space is required to run both nodes. Keep in mind that the storage requirements will grow over time.
```
It could be useful to mention, instead of "both nodes", that we're referring to run "a forger node for both, testnet and mainnet network", just to not get people confused on running "both rpc and forger node".

lander86 reviewed (approved):
Looks good to me. 
What about also mentioning that, if you delete the volume, you loose your wallet keys (that you should have safely stored somewhere anyway)? 
An in the "generate_keys" script I would also post a message that recommends to take a backup of the generated keys.

ysibirski commented on README.md:
```diff
@@ -1,124 +1,26 @@
 
----
-# Deployment
+This repository contains all the resources for deploying an EVM sidechain Node on mainnet or testnet. It can create either rpc nodes or forger nodes.
```
How about his better: "This repository contains all the resources for deploying a forger or rpc EVM sidechain node on mainnet or testnet."

ysibirski commented on compose_files/docker-compose-forger.yml:
```diff
@@ -16,12 +17,13 @@ networks:
     networks:
       evmapp_network:
-        ipv4_address: 10.10.50.10
+        ipv4_address: ${ZEND_IP_ADDRESS}
```
I think you should try use double quotes if works as expected here as well to follow your pattern before(all bash env vars enclosed in double quotes)

ysibirski commented on compose_files/docker-compose-forger.yml:
```diff
@@ -45,21 +47,22 @@ services:
     networks:
       evmapp_network:
-        ipv4_address: ${SCNODE_IP_ADDRESS}
+        ipv4_address: ${EVMAPP_IP_ADDRESS}
```
I think you should try use double quotes if works as expected here as well to follow your pattern before(all bash env vars enclosed in double quotes)

ysibirski commented on docs/FORGER.md:
```diff
@@ -0,0 +1,115 @@
+
+--- 
+
+As this is a forger node, it requires a zend node to run as well. Syncing a zend node from scratch may take a few hours, 
```
How about this: "Forger node requires zend node to be running as well."

ysibirski commented on docs/FORGER.md:
```diff
@@ -0,0 +1,115 @@
+--- 
+
+As this is a forger node, it requires a zend node to run as well. Syncing a zend node from scratch may take a few hours, 
+so a seed file can be used to speed up the process.
```
Replace 'so' with `therefore'

ysibirski commented on docs/FORGER.md:
```diff
@@ -0,0 +1,115 @@
+
+- `ZEN_FORCE_RESEED`: This variable is set to `false` by default, but it can be set manually to `true` if forcing the reseed of the zend node is required.
+
+Once the zend node has been synced, the following times the stack is started there is no need to seed the zend node again.
```
```suggestion
After the zend node has been synced, there is no need to seed it again when the stack is started subsequently.
```

ysibirski commented on docs/FORGER.md:
```diff
@@ -0,0 +1,115 @@
+
+The _seed.sh_ script will generate a **.seed.complete** file in the node's data folder, and it will prevent the seed script to be run again.
+
+Additionally, setting the `ZEN_USE_SEED_FILE` variable to false in the .env file will prevent the seed process to be run again.
```
`false` to match your previous style

ysibirski commented on docs/FORGER.md:
```diff
@@ -0,0 +1,115 @@
+
+Additionally, setting the `ZEN_USE_SEED_FILE` variable to false in the .env file will prevent the seed process to be run again.
+
+If migrating from previous versions of these project and your zend node is already synced set the `ZEN_USE_SEED_FILE` variable to false in the .env file.
```
`false`

ysibirski commented on docs/FORGER.md:
```diff
@@ -0,0 +1,115 @@
+
+If migrating from previous versions of these project and your zend node is already synced set the `ZEN_USE_SEED_FILE` variable to false in the .env file.
+
+If reseed of the zend node is required, set the `ZEN_FORCE_RESEED` variable to true in the .env file, and restart the stack.
```
`true`

ysibirski commented on docs/FORGER.md:
````diff
@@ -0,0 +1,115 @@
+    cd deployments/forger/[eon|gobi] && docker compose up -d zend
+    ```
+
+3. Verify if the zend node is fully synced you can run the following command and compare the output with the current block height in the mainchain: https://explorer.horizen.io or https://explorer-testnet.horizen.io:
````
```suggestion
3. Verify if zend node is fully synced by running the following command and comparing the output with the current block height in the mainchain: https://explorer.horizen.io or https://explorer-testnet.horizen.io:
```

ysibirski commented on docs/FORGER.md:
````diff
@@ -0,0 +1,115 @@
+    ```
+   **STORE THESE VALUES IN A SAFE PLACE. THESE VALUES WILL BE REQUIRED IN THE STAKING PROCESS**
+
+   Verify that the keys were generated correctly, run the following command:
````
```suggestion
   Verify that the keys were generated correctly by running the following command:
```

ysibirski commented on env/.env.rpc.eon.template:
```diff
@@ -0,0 +1,62 @@
+# Sidechain config
+SCNODE_ROLE=rpc
+SCNODE_FORGER_ENABLED=false
+SCNODE_FORGER_MAXCONNECTIONS=
```
I think you can get rid of this VAR for RPC node

ysibirski commented on env/.env.rpc.gobi.template:
```diff
@@ -0,0 +1,61 @@
+# Sidechain config
+SCNODE_ROLE=rpc
+SCNODE_FORGER_ENABLED=false
+SCNODE_FORGER_MAXCONNECTIONS=
```
Same here

ysibirski commented on scripts/forger/seed.sh:
```diff
@@ -0,0 +1,148 @@
+
+set -eEuo pipefail
+
+# This script is used to download and extract the blockchain data from a seed node.
```
from a seed file vs node?

ysibirski commented on scripts/forger/seed.sh:
```diff
@@ -0,0 +1,148 @@
+
+# Check if script needs to be run
+if [ "${FORCE_RESEED:-false}" = "true" ]; then
+  echo "FORCE_RESEED is true. Removing seed lock and complete files."
```
is set to true

ysibirski replied:
```suggestion
  echo "FORCE_RESEED is set to true. Removing seed lock file(s) and preparing data directory for the seed process."
```

ysibirski commented on scripts/forger/seed.sh:
```diff
@@ -0,0 +1,148 @@
+  echo "'.seed.complete' file exists. Skipping seed step."
+  exit 0
+elif [ "${USE_SEED_FILE:-false}" = "false" ]; then
+  echo "USE_SEED_FILE is false. Skipping seed step."
```
is set to false

ysibirski commented on scripts/forger/seed.sh:
```diff
@@ -0,0 +1,148 @@
+  touch "${datadir}/.seed.complete"
+  exit 0
+elif [ -d "${datadir}/blocks" ] && [ "$(ls -A "${datadir}"/blocks)" ]; then
+  echo "'blocks' directory exist and is not empty. Skipping seed step."
```
exists

ysibirski commented on scripts/forger/seed.sh:
```diff
@@ -0,0 +1,148 @@
+  touch "${datadir}/.seed.complete"
+  exit 0
+elif [ -d "${datadir}/chainstate" ] && [ "$(ls -A "${datadir}"/chainstate)" ]; then
+  echo "'chainstate' directory exist and is not empty. Skipping seed step."
```
exists

ysibirski commented on scripts/forger/seed.sh:
```diff
@@ -0,0 +1,148 @@
+    if ! aria2c -x 1 -s 1 -d "${datadir}" "${SEED_TAR_GZ_URL}" 2>/dev/null; then
+      echo "Download failed on attempt $i"
+      if ((i == retries)); then
+        echo "All download attempts failed. Exiting."
```
Error: all download attempts have failed. Exiting.

ysibirski commented on scripts/forger/seed.sh:
```diff
@@ -0,0 +1,148 @@
+
+  # Download loop
+  for ((i = 1; i <= retries; i++)); do
+    echo "Download attempt $i of $retries"
```
Seed file download attempt ....

ysibirski commented on scripts/forger/seed.sh:
```diff
@@ -0,0 +1,148 @@
+
+  # Extraction loop
+  for ((i = 1; i <= retries; i++)); do
+    echo "Extraction attempt $i of $retries"
```
Seed file extraction ...

ysibirski commented on scripts/forger/seed.sh:
```diff
@@ -0,0 +1,148 @@
+      rm -rf "${datadir}/chainstate"/*
+
+      if ((i == retries)); then
+        echo "All extraction attempts failed. Removing directories and exiting."
```
Error: all extraction attempts have failed. Cleaning up and exiting.

ysibirski commented on scripts/forger/seed.sh:
```diff
@@ -0,0 +1,148 @@
+    if ! curl -L "${SEED_TAR_GZ_URL}" | tar -xzf - -C "${datadir}"; then
+      echo "Import failed on attempt $i"
+      if ((i == retries)); then
+        echo "All import attempts failed. Exiting."
```
Error: all import attempts have failed. Exiting.

ysibirski commented on scripts/forger/seed.sh:
```diff
@@ -0,0 +1,148 @@
+if [ "${datadir_available_space_bytes}" -ge ${total_size} ]; then
+  download_and_extract_seed "${datadir}" "${SEED_TAR_GZ_URL}"
+else
+  import_seed "${datadir}" "${SEED_TAR_GZ_URL}"
```
Tell me more about this. Import seed will extract from the remote, correct? Why do we want download and extract then for?

otoumas replied:
The idea is to split the download and extract processes if there is enough space, this way download process should be faster and not depend on the extraction. It also makes it easier to clean up afterwards.

ysibirski commented on scripts/init.sh:
```diff
@@ -1,184 +1,157 @@
+
+  if [ "${role_value}" = "forger" ]; then
+    echo -e "\nYou are setting up a forger node with a zend node. Zend node may require a few hours to sync.\n"
+    read -rp "Do you want to download and use a seed file to speed up the process? ('yes' or 'no'): " use_seed_answer
```
"speed up zend syncing process"

ysibirski commented on scripts/init.sh:
```diff
@@ -1,184 +1,157 @@
+echo -e "\n\033[1m=== Creating symlink to compose file ===\033[0m"
+COMPOSE_FILE="${ROOT_DIR}/compose_files/docker-compose-${role_value}.yml"
+SYMLINK_COMPOSE_FILE="${DEPLOYMENT_DIR}/docker-compose.yml"
+if ! [ -L "${SYMLINK_COMPOSE_FILE}" ]; then
+  ln -sf "${COMPOSE_FILE}" "${SYMLINK_COMPOSE_FILE}"
+fi
```
You do not even need to check if it is exist since `ln -sf` is forcing regardless

ysibirski commented on scripts/init.sh:
```diff
@@ -1,184 +1,157 @@
 
-    echo "Thanks for your interest in EON, your configuration is incomplete. Application will not be able to start. Please contact Horizen to fix this issue. Exiting ..."
-    exit 1
+  echo -e "\n\033[1m=== Creating symlink to seed file ===\033[0m"
```
I would drop this message completely in case people will not be choosing `seed` method but seeing this message can be misleading.

If you keep it then change `seed file` to `seed.sh script file`

ysibirski commented on scripts/init.sh:
```diff
@@ -1,184 +1,157 @@
+  echo -e "\n\033[1m=== Creating symlink to seed file ===\033[0m"
+  SEED_FILE="${ROOT_DIR}/scripts/forger/seed.sh"
+  SYMLINK_SEED_FILE="${DEPLOYMENT_DIR}/scripts/seed.sh"
+  if ! [ -L "${SYMLINK_SEED_FILE}" ]; then
+    mkdir -p "${DEPLOYMENT_DIR}/scripts"
+    ln -sf "${SEED_FILE}" "${SYMLINK_SEED_FILE}"
```
Same here if you are `ln -sf` does not matter if it exist or not

ysibirski commented on scripts/init.sh:
```diff
@@ -1,184 +1,157 @@
-# Starting evmapp
-######
-cd "${ROOT_DIR}"
+  echo -e "\n\033[1m=== Project has been setup correctly for ${role_value} and ${network_value} ===\033[0m"
```
initialized vs setup

ysibirski commented on scripts/init.sh:
```diff
@@ -1,184 +1,157 @@
-  docker update --restart=no "${CONTAINER_NAME}" &>/dev/null
-  $COMPOSE_CMD -f ${COMPOSE_FILE} exec "${CONTAINER_NAME}" gosu user curl -s -X POST "http://127.0.0.1:${SCNODE_REST_PORT}/node/stop" -H "accept: application/json" -H 'Content-Type: application/json' &>/dev/null
-  sleep 5
+  echo -e "1. Run first the zend node:"
```
```suggestion
  echo -e "1. First, run zend node:"
```

ysibirski commented on scripts/init.sh:
```diff
@@ -1,184 +1,157 @@
 
-# Making sure scnode is up and running
-scnode_start_check
+  echo -e "2. Verify height of the explorer: ${EXPLORER_URL} matches your node height:"
```
```suggestion
  echo -e "2. Verify your node's block height matches against the public explorer: ${EXPLORER_URL}:"
```

ysibirski commented on scripts/init.sh:
```diff
@@ -1,184 +1,157 @@
+  echo -e "\n\033[1m===========================\033[0m\n"
+else
+
+  echo -e "\n\033[1m=== Project has been setup correctly for ${role_value} and ${network_value} ===\033[0m"
```
initialized vs setup

ysibirski commented on scripts/utils.sh:
```diff
@@ -1,92 +1,136 @@
-}
+  command -v docker &>/dev/null || fn_die "${FUNCNAME[0]} error: 'docker' is required to run this script, see installation instructions at 'https://docs.docker.com/engine/install/ubuntu/#install-using-the-repository'."
+
+  (docker-compose version 2>&1 | grep -q v2 || docker compose version 2>&1 | grep -q v2) || fn_die "${FUNCNAME[0]} error: 'docker-compose' or 'docker compose' v2 is required to run this script, see installation instructions at 'https://docs.docker.com/compose/install/other/'."
```
change `https://docs.docker.com/compose/install/other/` to `https://docs.docker.com/compose/install/`

From this new page it is also saying installing standalone compose is NOT recommended anymore. I wonder if we should drop supporting it at all

otoumas replied:
well, I would be glad to replace support in these scripts.

@lander86 @cronicc ??

lander86 replied:
As already agreed, let's proceed by removing support to the standalone compose.

ysibirski reviewed (commented):
Please read and perhaps consider some comments.

cronicc commented on env/.env.forger.eon.template:
```diff
@@ -66,7 +70,7 @@ SCNODE_GRP_ID=999
 SCNODE_WS_SERVER_PORT=9110
-SCNODE_WS_CLIENT_ENABLED=
+SCNODE_WS_CLIENT_ENABLED=true
 SCNODE_WS_SERVER_ENABLED=true
```
I think we should disable the WS Server by default on Forger Nodes. For optimal performance Forgers shouldn't be used as RPC/WS nodes.

cronicc commented on docs/FORGER.md:
```diff
@@ -0,0 +1,115 @@
+# Forger Node
```
Comment for file deployments/.gitignore. I think we need to add 
```
*
!.gitignore
```
to be sure that a deployment doesn't cause unstaged changes in git.

cronicc commented on scripts/forger/seed.sh:
```diff
@@ -0,0 +1,148 @@
+  for ((i = 1; i <= retries; i++)); do
+    echo "Download attempt $i of $retries"
+
+    if ! aria2c -x 1 -s 1 -d "${datadir}" "${SEED_TAR_GZ_URL}" 2>/dev/null; then
```
The advantage of aria2c is that it can greatly speed up DLs by using up to 16 parallel connections with `-x16 -s16`. Using `-x1 -s1` provides no advantage over using curl. Depending on the server we're downloading from this doesn't always work reliably. Cloudflare non-enterprise for example only allows up to 2 connections.

I wrote logic a few years ago in zen-fetch-params to handle these cases in addition to resume support of partially downloaded files. https://github.com/HorizenOfficial/zen/blob/f0c80a6853d2e77b39d019ac62fe74210e15a0aa/zcutil/fetch-params.sh#L31-L71

lander86 commented:
In generate_keys script there is the utils.sh dependency with the wrong path (there were no "forger" directory before the changes). Please fix it at line 6 !

Commit:
Address code review comments.
Split seed process, download and extarct is done now before running zend node, and can be done either using a script or manually.

cronicc commented on scripts/forger/seed/download_seed.sh:
```diff
@@ -0,0 +1,115 @@
+  echo "Seed file size: ${content_length} bytes"
+  echo "Minimal required storage size: ${extended_content_length} bytes"
+  echo "Recommended storage size: ${total_size} bytes"
+} >>"${seed_dir}/seed.log"
```
Non-blocking, directing both stdout and stderr to the logfile and stdout would provide more information on progress to the user.

`2>&1 | tee -a "${seed_dir}/seed.log"` instead of `>>"${seed_dir}/seed.log"` everywhere we're redirecting.

otoumas replied:
Done.

Commit:
Improve download_seed.sh logging

ysibirski commented on scripts/forger/seed/download_seed.sh:
```diff
@@ -24,23 +24,24 @@ extended_content_length=$((content_length + content_length / 10))
   echo "Minimal required storage size: ${extended_content_length} bytes"
   echo "Recommended storage size: ${total_size} bytes"
-} >>"${seed_dir}/seed.log"
+} | tee -a "${seed_dir}/seed.log"
```
I think you are missing stderr to stdout redirection everywhere before the pipe `2>&1`

ysibirski commented on scripts/forger/seed/download_seed.sh:
```diff
@@ -56,9 +57,9 @@ if [ "${seed_dir_available_space_bytes}" -ge ${total_size} ]; then
     --download-result=full \
     --summary-interval=10 \
-    "${remote_file}" >>"${seed_dir}/seed.log" ||
+    "${remote_file}" | tee -a "${seed_dir}/seed.log" ||
```
Everywhere where you do NOT use just ECHO command you want `2>&1` stderr to stdout redirection in case the command fails and you will save its error to log file as well.
To be honest you can add `2>&1` to all you commands even echo. It will not hurt

ysibirski commented on scripts/forger/seed/download_seed.sh:
```diff
@@ -70,46 +71,46 @@ if [ "${seed_dir_available_space_bytes}" -ge ${total_size} ]; then
 
-    if ! curl -L "${remote_file}" | tar -xzf - -C "${seed_dir}" >>"${seed_dir}/seed.log"; then
-      echo "Seed file extraction failed on attempt $i" >>"${seed_dir}/seed.log"
+    if ! curl -L "${remote_file}" | tar -xzf - -C "${seed_dir}" | tee -a "${seed_dir}/seed.log"; then
```
Same here.

ysibirski reviewed (changes requested):
Some changes are required

ysibirski commented on scripts/forger/seed/download_seed.sh:
```diff
@@ -0,0 +1,115 @@
+env_file="${deployment_dir}/.env"
+
+# shellcheck source=../.env
+source "${env_file}"
```
See if you want to check first if env file exist and exit with the message if does not

ysibirski commented on scripts/forger/seed/seed.sh:
```diff
@@ -0,0 +1,62 @@
+# Main
+echo "Starting the seed process..."
+
+# Check if blocks_dir or chainstate_dir does not exist or is empty
+if [ ! -d "${blocks_dir}" ] || [ ! "$(ls -A "${blocks_dir}")" ] || [ ! -d "${chainstate_dir}" ] || [ ! "$(ls -A "${chainstate_dir}")" ]; then
+  echo "Either ${blocks_dir} or ${chainstate_dir} directories do not exist or are empty. Nothing to seed. Exiting script."
+  exit 0
+fi
```
Isn't it that if directories are empty that means YES lets seed vs nothing to seed? :shrug:

otoumas replied:
Not in this case, here the check is to find out if there is anything to seed, so these directories are the blocks and chainstate that come form the tar file and need to replace the blocks and chainstate of the node, so if they don't exist don't do anything

ysibirski replied:
Does it mean it checks if the seed tar file is kind of broken in this case?

otoumas replied:
No, it checks if there is a seed to use. The logic is:
- when node starts it always runs the seed script
- If there is NOT a seed (blocks and chainstate directories exists in seed directory and are not empty) then it will exit the script because there is no seed.
- If there is a seed (blocks and chainstate directories exists in seed directory and are not empty) then:
- if .seed.complete file exists: exit the seed script
- If blocks or chainstate of the node's datadir exists exit script.
- If any of these exists and there is a seed and you want to force the re seed there is an env var to control it.

So the first time you run the zend node you may want to seed it, so you place the blocks and chainstate dirs in the seed dir.
But if you stop your nodes and start them again the seed folder will be empty because you don't want to reseed. 
This is what that first if is checking, is there a seed?

ysibirski replied:
By `place the blocks and chainstate dirs in the seed dir` you mean like manually or by some other script/function if someone would populate those `/seed/{blocks,chainstate} dirs` and then run `seed.sh` ?

otoumas replied:
Both. Process is described in the FORGER.md file. Now you can either use download_seed.sh file or do it manually.
No need to run seed.sh file as it will be run by the node as it is the custom script.

Commit:
Applied stderr to stdout redirects

ysibirski reviewed (approved):
LGTM

Commit:
Add SCNODE_REST_PASSWORD to compose files

Commit:
Remove comment on env.templates

Commit:
Modify zend ZEN_OPTS env var
